src/btree/bt_read.c

```diff
@@ -105,28 +105,24 @@ __page_read(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags)
      */
     WT_CLEAR(tmp);
 
-    /*
-     * Attempt to set the state to WT_REF_READING for normal reads, or WT_REF_LOCKED, for deleted
-     * pages. The difference is that checkpoints can skip over clean pages that are being read into
-     * cache, but need to wait for deletes to be resolved (in order for checkpoint to write the
-     * correct version of the page).
-     *
-     * If successful, we've won the race, read the page.
-     */
+    /* Lock the WT_REF. */
     switch (previous_state = ref->state) {
     case WT_REF_DISK:
-        new_state = WT_REF_READING;
-        break;
     case WT_REF_DELETED:
-        new_state = WT_REF_LOCKED;
-        break;
+        if (WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
+            break;
+        return (0);
     default:
         return (0);
     }
-    if (!WT_REF_CAS_STATE(session, ref, previous_state, new_state))
-        return (0);
 
-    final_state = WT_REF_MEM;
+    /*
+     * Set the WT_REF_READING flag for normal reads. Checkpoints can skip over clean pages being
+     * read into cache, but need to wait for deletes to be resolved (in order for checkpoint to
+     * write the correct version of the page).
+     */
+    if (previous_state == WT_REF_DISK)
+        F_SET(ref, WT_REF_READING);
```

Comment:
I'm still reading through this, but what stops two threads trying to read the same page into cache?

Reply:
Oh, I see, ignore me.  The diff formatted oddly and made me miss the fall through.

Reply:
> I'm still reading through this, but what stops two threads trying to read the same page into cache?

The `WT_REF` state is set to `WT_REF_LOCKED`.